llvm/lib/Target/X86/X86ISelLowering.cpp

```diff
@@ -1970,20 +1970,24 @@ X86TargetLowering::X86TargetLowering(const X86TargetMachine &TM,
     }
 
     if (Subtarget.hasVBMI2()) {
-      for (auto VT : { MVT::v8i16, MVT::v4i32, MVT::v2i64,
-                       MVT::v16i16, MVT::v8i32, MVT::v4i64,
-                       MVT::v32i16, MVT::v16i32, MVT::v8i64 }) {
+      for (auto VT : {MVT::v32i16, MVT::v16i32, MVT::v8i64}) {
         setOperationAction(ISD::FSHL, VT, Custom);
         setOperationAction(ISD::FSHR, VT, Custom);
       }
 
       setOperationAction(ISD::ROTL, MVT::v32i16, Custom);
-      setOperationAction(ISD::ROTR, MVT::v8i16,  Custom);
-      setOperationAction(ISD::ROTR, MVT::v16i16, Custom);
       setOperationAction(ISD::ROTR, MVT::v32i16, Custom);
     }
   }// useAVX512Regs
 
+  if (Subtarget.hasVBMI2()) {
```

Comment:
Now `setOperationAction` is called without `!Subtarget.useSoftFloat()`. I've just realized that we have this predicate on integer operations. I don't know the reasons why it is so. I assume it is used as a vector register availability predicate. If so, then we need it here as well

Reply:
Thanks for the reminder!
I think `useSoftFloat` only matters when calling to `addRegisterClass`. We don't have any chance to lower the opcode with a legal type.
Anyway, I think adding `useSoftFloat` matches the conversions better, I'll add it in another patch.

Reply:
https://github.com/llvm/llvm-project/commit/006f264b0343d0a27a95854e4f3aa610e5dac3bb